File: 2048GAME/python/ai.py
"""
ai.py — AIEngine: Expectimax AI for 2048 (Python)
"""
import random
import math
from game import GameBoard


class AIEngine:
    DEPTHS = {'easy': 0, 'normal': 3, 'hard': 5, 'extreme': 7}

    def __init__(self, difficulty='normal'):
        self.difficulty = difficulty

    def best_move(self, board):
        if self.difficulty == 'easy':
            return self._random_move(board)
        depth = self.DEPTHS[self.difficulty]
        return self._expectimax(board, depth)

    # ---- Random ----
    def _random_move(self, board):
        moves = board.valid_moves()
        return random.choice(moves) if moves else -1

    # ---- Expectimax ----
    def _expectimax(self, board, depth):
        moves = board.valid_moves()
        if not moves:
            return -1
        best_score, best_dir = float('-inf'), moves[0]
        for d in moves:
            b = board.clone()
            moved, _ = b.move(d, sim=True)
            if not moved:
                continue
            s = self._expect_node(b, depth - 1)
            if s > best_score:
                best_score, best_dir = s, d
        return best_dir

    def _expect_node(self, board, depth):
        if depth == 0 or board.over:
            return self._heuristic(board)
        cells = board.empty_cells()
        if not cells:
            return self._heuristic(board)
        sample = cells if depth <= 2 else random.sample(cells, min(4, len(cells)))
        score = 0.0
        for (r, c) in sample:
            for val, prob in ((2, 0.9), (4, 0.1)):
                b = board.clone()
                b.grid[r][c] = val
                score += prob * self._max_node(b, depth - 1)
        return score / len(sample)

    def _max_node(self, board, depth):
        if depth == 0 or board.over:
            return self._heuristic(board)
        moves = board.valid_moves()
        if not moves:
            return self._heuristic(board)
        best = float('-inf')
        for d in moves:
            b = board.clone()
            moved, _ = b.move(d, sim=True)
            if not moved:
                continue
            s = self._expect_node(b, depth - 1)
            if s > best:
                best = s
        return best

    # ---- Heuristic ----
    def _heuristic(self, board):
        g = board.grid

        def lg(v):
            return math.log2(v) if v > 0 else 0.0

        empty = sum(1 for row in g for v in row if v == 0)

        mono = 0.0
        for r in range(4):
            inc = dec = 0.0
            for c in range(1, 4):
                d = lg(g[r][c]) - lg(g[r][c - 1])
                if d > 0:
                    inc += d
                else:
                    dec -= d
            mono -= min(inc, dec)
        for c in range(4):
            inc = dec = 0.0
            for r in range(1, 4):
                d = lg(g[r][c]) - lg(g[r - 1][c])
                if d > 0:
                    inc += d
                else:
                    dec -= d
            mono -= min(inc, dec)

        smooth = 0.0
        for r in range(4):
            for c in range(4):
                if g[r][c]:
                    if c < 3 and g[r][c + 1]:
                        smooth -= abs(lg(g[r][c]) - lg(g[r][c + 1]))
                    if r < 3 and g[r + 1][c]:
                        smooth -= abs(lg(g[r][c]) - lg(g[r + 1][c]))

        max_val = board.max_tile()
        corners = [g[0][0], g[0][3], g[3][0], g[3][3]]
        corner_bonus = lg(max_val) * 2.0 if max_val in corners else 0.0

        merges = 0
        for r in range(4):
            for c in range(4):
                if c < 3 and g[r][c] and g[r][c] == g[r][c + 1]:
                    merges += 1
                if r < 3 and g[r][c] and g[r][c] == g[r + 1][c]:
                    merges += 1

        return (empty * 270 + mono * 47 + smooth * 0.1 +
                corner_bonus * 100 + merges * 700 + board.score * 0.1)

File: 2048GAME/nodejs/ai.js
'use strict';
// ai.js — AIEngine: Expectimax AI for 2048 (Node.js CLI)
const GameBoard = require('./game');

class AIEngine {
  constructor(difficulty = 'normal') {
    this.difficulty = difficulty;
    this.depths = { easy: 0, normal: 3, hard: 5, extreme: 7 };
  }

  bestMove(board) {
    if (this.difficulty === 'easy') return this._randomMove(board);
    return this._expectimax(board, this.depths[this.difficulty] || 3);
  }

  _randomMove(board) {
    const m = board.validMoves();
    return m.length ? m[Math.floor(Math.random() * m.length)] : -1;
  }

  _expectimax(board, depth) {
    const moves = board.validMoves();
    if (!moves.length) return -1;
    let bestScore = -Infinity, bestDir = moves[0];
    for (const d of moves) {
      const b = board.clone();
      if (!b.move(d, true).moved) continue;
      const s = this._expectNode(b, depth - 1);
      if (s > bestScore) { bestScore = s; bestDir = d; }
    }
    return bestDir;
  }

  _expectNode(board, depth) {
    if (depth === 0 || board.over) return this._heuristic(board);
    const cells = board.emptyCells();
    if (!cells.length) return this._heuristic(board);
    const sample = depth <= 2 ? cells : this._sample(cells, 4);
    let score = 0;
    for (const [r, c] of sample)
      for (const [val, prob] of [[2, 0.9], [4, 0.1]]) {
        const b = board.clone(); b.grid[r][c] = val;
        score += prob * this._maxNode(b, depth - 1);
      }
    return score / sample.length;
  }

  _maxNode(board, depth) {
    if (depth === 0 || board.over) return this._heuristic(board);
    const moves = board.validMoves();
    if (!moves.length) return this._heuristic(board);
    let best = -Infinity;
    for (const d of moves) {
      const b = board.clone();
      if (!b.move(d, true).moved) continue;
      const s = this._expectNode(b, depth - 1);
      if (s > best) best = s;
    }
    return best;
  }

  _heuristic(board) {
    const g = board.grid;
    const lg = v => v > 0 ? Math.log2(v) : 0;
    const empty = g.flat().filter(v => !v).length;

    let mono = 0;
    for (let r = 0; r < 4; r++) {
      let inc = 0, dec = 0;
      for (let c = 1; c < 4; c++) {
        const d = lg(g[r][c]) - lg(g[r][c-1]);
        d > 0 ? (inc += d) : (dec -= d);
      }
      mono -= Math.min(inc, dec);
    }
    for (let c = 0; c < 4; c++) {
      let inc = 0, dec = 0;
      for (let r = 1; r < 4; r++) {
        const d = lg(g[r][c]) - lg(g[r-1][c]);
        d > 0 ? (inc += d) : (dec -= d);
      }
      mono -= Math.min(inc, dec);
    }

    let smooth = 0;
    for (let r = 0; r < 4; r++)
      for (let c = 0; c < 4; c++) {
        if (!g[r][c]) continue;
        if (c < 3 && g[r][c+1]) smooth -= Math.abs(lg(g[r][c]) - lg(g[r][c+1]));
        if (r < 3 && g[r+1][c]) smooth -= Math.abs(lg(g[r][c]) - lg(g[r+1][c]));
      }

    const maxVal = board.maxTile();
    const corners = [g[0][0], g[0][3], g[3][0], g[3][3]];
    const cornerBonus = corners.includes(maxVal) ? lg(maxVal) * 2 : 0;

    let merges = 0;
    for (let r = 0; r < 4; r++)
      for (let c = 0; c < 4; c++) {
        if (c < 3 && g[r][c] && g[r][c] === g[r][c+1]) merges++;
        if (r < 3 && g[r][c] && g[r][c] === g[r+1][c]) merges++;
      }

    return empty*270 + mono*47 + smooth*0.1 + cornerBonus*100 + merges*700 + board.score*0.1;
  }

  _sample(cells, n) {
    return [...cells].sort(() => Math.random() - 0.5).slice(0, n);
  }
}

module.exports = AIEngine;
